opt: refactor and fix building of RETURNING expressions

mgartner opened the pull request:
#### opt: refactor building of RETURNING expressions and scopes

The logic for constructing `RETURNING` clauses has been refactored. The
logic is almost identical to what it was before. The only difference is
that we now always build the returning scopes for inserts before
building check constraints and we only build those scopes once even if
RLS is enabled. The latter is beneficial for reducing the max column ID
of query plans, helping to avoid the "slow path" of column sets which
require heap allocations - a column set with a column ID greater than
128 hits this slow path.

Release note: None

#### opt: fix application of RLS SELECT policies for INSERTs

It was previously assumed that a `RETURNING` clause of an `INSERT` could
only reference columns of the target table. Therefore, we simply checked
whether or not any columns were referenced in the `RETURNING` clause to
determine if RLS SELECT policies should be applied as check constraints.

However, our assumption was incorrect. If the `INSERT` is within a UDF
or stored procedure, the `RETURNING` clause can reference parameters.
These references should not cause RLS SELECT policies to be applied.
This commit fixes this.

Fixes #146470

Release note: None

mgartner committed:
opt: refactor building of RETURNING expressions and scopes

The logic for constructing `RETURNING` clauses has been refactored. The
logic is almost identical to what it was before. The only difference is
that we now always build the returning scopes for inserts before
building check constraints and we only build those scopes once even if
RLS is enabled. The latter is beneficial for reducing the max column ID
of query plans, helping to avoid the "slow path" of column sets which
require heap allocations - a column set with a column ID greater than
128 hits this slow path.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146394)
<!-- Reviewable:end -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Much cleaner. I like that it backed out my updates to xform/testdata/external/tpce*. I felt uneasy about that change.

> Reviewed 7 of 7 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146394#-:-OPp5bBW5LzhYGtgewq4:bygjoyg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

mgartner commented:
> I like that it backed out my updates to xform/testdata/external/tpce*. I felt uneasy about that change.

I think it backed out some, but it made some new changes too. It all looks fine to me though.

mgartner commented:
It turns out the changes in the second commit are incorrect. A `RETURNING` clause can reference columns not in the target table, such as UDF/SP parameters. Based on this, I found a minor RLS bug, detailed in #146470.

So I'll revert the second commit and add a new one to fix #146470.

mgartner committed:
opt: fix application of RLS SELECT policies for INSERTs

It was previously assumed that a `RETURNING` clause of an `INSERT` could
only reference columns of the target table. Therefore, we simply checked
whether or not any columns were referenced in the `RETURNING` clause to
determine if RLS SELECT policies should be applied as check constraints.

However, our assumption was incorrect. If the `INSERT` is within a UDF
or stored procedure, the `RETURNING` clause can reference parameters.
These references should not cause RLS SELECT policies to be applied.
This commit fixes this.

Fixes #146470

Release note: None

mgartner commented:
@spilchen please re-review, specifically the second commit which is new.

spilchen reviewed (approved):
Good find with latest issue. 
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146394#-:-OPr_1Ky3okD1MD_KiLe:b-sdo2al)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
